### DISCUSSION TITLE
Welcome to Verse3 Discussions!

### DISCUSSION THREAD

**Discussion by pm-Architect** [iiterate-de/Verse3#9] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by RantNRave31** [comment]
Im very interested.  it did not build with vs2022.. 19?  its the "setup projects" that are incompatible.
I'm attempting to figure out how to run from build directory as well.

i've tried visualscriptingengine, pure data, and flowgrithm, which i love.
i love the blue print engines in both blender and Unreal, .. oh, and unity.

i'm interested to see which way you went.   from the libraries, it appears you are headed in the right direction.

I, will downgrade, i.e. install VS 2015, and try it's installer version so wee what version the setup/installer you are using.?

I'll keep you updated.  Imagine if you will, a man, gasping in the heat of the desert, but "drooling" on what i've seen looking though your code.  ?   REST??? and visual scripting?  interop?  do you mean, import a library, and generate nodes form it's objects?  man? if that's true, i'm going to die.

i've looked at a number and am excited to see which way you went?  did you tie in the presentation to the action?  or did you script the action? or, do you compile the action with LLVM?  hmmmm....   i bet your graphs generate an AST? like one we can use LLVM to compile and run?

we'll see.  from the list of libraries, you been very busy and VERY thorough.

what will this old man find, as he peruses your source.  i already know there is one bright light behind this.  hmmm?  ;-)  have a GREAT day.  if you get back, let me knwo what version VS to build with.  and what setup extension should i download?

Regards,
Donnie G

**Reply by RantNRave31** [reply]
P.S.  Yo, master Yoda, please assist me getting this up, and please, teach me how to add components for neural networks, bayesian networks, so i can play with classifiers and convolution networks ..   yeah.. right.  dude.   i shouldn't have smoke that joint before writting this.  please excuse the ... laziness of informal grammer.  i do much better when coding lol.  later.

**Reply by RantNRave31** [reply]
AHAHAHAHAH!!! It friggin WORKS.  man!  outstanding job.
i had to:
to build the project when you can't find the msi or build the setup.

1. In $AppData "Roaming", Create an Verse3 folder with sub-directories (CanvasExports and Libraries)
2. Build the AddIn Libraries (no error now, they can copy into libraries.
3. Build the Verse 3 (it fails, no Monaco folder.
4. Copy the Monaco folder to the BUILD, in my case under ..\debud\.. folder
5. Build the Verse 3 (it now builds)
6. Run the Verse3!!! .  yes indeed.  
7. Applaud the Author!  You da MAN brother.  this is nice.
{CLAP CLAP CLAP}
cue the music.
fade to black.

**Comment by pm-Architect** [comment]
Hi @RantNRave31 !
I'm glad that you got it up and running by yourself! 😄 I really appreciate your kind words, thanks!
You shouldn't need to downgrade VS to get it running. Just make sure all the dependencies are set up correctly - The project is built using WPF.
Also, it's still VERY buggy and not complete by far. I'm not actively developing this version anymore.
I'm now working on the 4th iteration in an effort to build a powerful visual programming language. This new version, called decode, is built to WebAssembly using the Blazor framework.
But now that you've got the project up and running, I'm really curious to get more feedback from you, especially from a developer's perspective. Looking forward to what you do with it! 🚀

**Reply by RantNRave31** [reply]
Sir!  i'm liking this very much.  If, you were to add a plugin system like "Pure Data", you can immediately pull Audio components in and "hear" your results.  as you did with shapes..  man.. i so much want to merge.  ...   i'm already thinkin.. i'm going to take it apart and put it back together to understand your design choices and increase my knowledge of your interface.

you have the beginnings, the core, of a full fledged C.A.S.E system that will be light years ahead of Sparx Enterprise Architect.

constructive criticism:  controls contents are not dense.. i.e lot of wasted space.  I'd like to see A "pallet" toolbar, with icons, so we can drag and drop, i.e. "paint" our diagrams.  i personally prefer titles at the top, but love your color schema.  I would like to be able to drag a line to a control instead of first clicking, releasing, clicking, and then dragging.  workflow change.

I'd like to see and explorer panel of the object tree.  yes indeed that would be nice.  but i've no funds for the labor cost ;-)  "dear santa, for christmas, i want an explorer panel."

I'd like to see you merge "Pure Data" components, but not GUI as yours kicks its ass.  then!! you can add sensors you your "connectors" and use signal theory to analyze your data flows... like a circuit tester but for outputing to an FFT or such.... yeah... this is kewl.

Dear Santa, may i please have an occiliscope plug in, and , and , and and......

ha, i could go on..... i will extend your product and let you know.  you da MAN!!!!!

your GUI design also facilitates DSL designers, and I'm interested to see if simple DSL like xml or activity designers can be implemented with your GUI.  to use Model to Model transforms.. and transform model into source.  or code.. whichever.  i like your approach, becuase i can simulate the program immediately.  

This leads me to suspect that you might be thinking this as well.  your output can fork to IML and LLVM can build an EXE or run on the fly.
dude.  that's ... exactly the right way to go man.   with the components you have added, i can infer certain "things" and am curious as heck to see what you do.  Keep up the good work.  i'm going to, most likely, adopt you.   hahahaha.  have a nice day.

seriously, the only thing you a really missing... is the plugin framework for chaining your graphs together.
Thanks for experience. and thanks for sharing your work with us.

**Reply by pm-Architect** [reply]
> If, you were to add a plugin system like "Pure Data"

A plugin system already exists! You can simply create a new C# project and class inheriting the abstract class `Comp`. Build it to a dll with the extension ".verse" and add it to `%AppData%\Roaming\Verse3\Libraries`. I recommend taking a look at [Grasshopper3D](https://www.grasshopper3d.com/) and plugins like [Firefly](http://www.fireflyexperiments.com/).

> you have the beginnings, the core, of a full fledged C.A.S.E system that will be light years ahead of Sparx Enterprise Architect.

I'm not sure what that is, but I'm really glad you like the project! 😄 

> constructive criticism: controls contents are not dense.. i.e lot of wasted space.

Well I have an excuse for the horrible buggy UI - WPF sucks 😆 That's part of the reason why I've moved on to developing decode. Since it's web-based, I'm trying to implement [FAST](https://www.fast.design/) UI elements to make my life a little easier as a solo developer. I post development updates on [LinkedIn](https://www.linkedin.com/in/praneet-mathur/) semi-regularly.

> but i've no funds for the labor cost ;-)

Spread the word!! That'd be the best way to invest in it :D

> your GUI design also facilitates DSL designers

I really believe the possibilities are endless. The goal here is to try and build a literal visual programming language. A language to interface with computers, without volumes of English-imitating "coded" text. To qualify as a programming language, it MUST be Turing complete and should be as powerful & capable as most modern programming languages. Beyond that, extensibility behaves somewhat like a modern npm or NuGet. Fun fact: with decode, the plugin installation system is [invisible](https://www.linkedin.com/feed/update/urn:li:activity:7064980795588788225).

> output can fork to IML and LLVM can build an EXE or run on the fly.

I am planning to explore the possibility to do this, but I was told to look at MSIL for achieving this in C#/.NET

I'm sure you'll enjoy decode more ;D Stay tuned!

**Reply by RantNRave31** [reply]
YES!... great.. plugin.  

you are certainly on the right track.   i'll keep using it.  i'll use the feed back form in app for bug reporting.

yeah.   this is gonna be very powerful.    Keep up the good work.  i'll check out plugins.

i'd prefer not going the browser path as i'm into high performance., Unreal, Unity, ISS45 POS, IBM ACE SurePOS, FreeBSD, Hadoop, Sparx, Apache Open Office, Apache OfBiz, NCR Connected Payments... stuff we can do a million dollars in ONE day across 33 lanes.  in a grocery store.      has to be fast.. 

i suppose if your browser is just the frontend to the server, then really, the way you have designed it, it's running on the server right?

i put my low performance low frequency access in the cloud, and high performance high access on site.

may this Blazor thing will work out?  I personally like WPF, but you may want to try GTK+?  same thing, presentation layer binding.

there is also MONO, opensouce Game engine you can use with presentation layer.

leave presentation to them, buisiness logic, compilation, and data logic to yourself.  ;-)

High Performance Business software should be "married" to High Performance Frontends.

if yours is fast enough to play a 3d game in a web browser page, you will the the KING!!!!!!!!!

NOTE:  i'm going to play with my new toy, your tool, and come up with some ideas.    muah ha hahahah <evil grin>

later bud.  have a great day.  and thanks for the look at your product.

no body wants to see my cash register LOL.. :-(

**Comment by RantNRave31** [comment]
Sir,
Mr. Praneet Mathur.

Nice work.
I've learned a lot.  I like a lot about it.  I am a Compiler Compiler and Code Generating Kind of Guy.  I use Enterprise Architect. and I write my own DSL Designer. 

I have a WPF Workspace Framework using an MVVM architecture that disconnects the Business Logic Layer Models from the Presentation Layer ViewModels and the Presentation Layer Views.  Shapes, Drag and Drop, Flows.  yours is better ;-) 

 I am attempting to add something similar to your InfiniteCanvas as a ModelView and have had some success.
I have finished learning your plugin model and am able to import your plugin Elements into my ToolPanels as Components.
DUDE!!!  i can see it, but i can't run it. LOL.  i wish to use your coreinterop server to provide Verse3 live simulations to my UML Activity Diagrams and BPMN process simulations.

They now load in my toolbox.  i'm taking a lesson from your "AddToCanvas_OnCall" to paint my shapes with your plugin elements.
drag from tool bar, drop onto editorModelView next.

Then, tie in to your interop server to executre right?  i'm not that far yet.  man.. simple and elegant.
i use a modified version of the Taste Grammar/AST/Parser to Generate compiled bytecode that an Interpreter runs.  but it has to compile.
YOU use sort of code server dude.
![image](https://github.com/pm-Architect/Verse3/assets/8115224/943684a0-2af6-4e24-b2fc-b7006317a04f)

**Reply by pm-Architect** [reply]
Really cool! Great to see that you're learning from my code 😄 🚀

**Reply by RantNRave31** [reply]
SIR!  dude ;-)  you are awsome.  you have done a wonderful job.  i've given up integrating.. and started learning to extend your product.  nice.    i love it.  i HOPE YOU DON'T MIND,....  but... instead of integrating mine, i ... decoupled your presentation from buisiness... and renamed models, modelviews, and viewmodels to better understand your architecture and to provide for WPF behaviors instead of your current method of wiring views together.  i prefer the view not have any knowledge of other views except through behaviors as it isolates display specifics from buisiness logic.  anyway... i converted all of your WinFORMS to WPF, merged MainWindow and Editor Windows.  THEN moved all references from the ModelViews and the ViewModels.
I refactored to single file per class so i could see.  and grouped objects by function.  after this sorting, and reduction... i've managed... heck.. to keep it working.. this is the 4th try but.. hey... persistance.

quick and dirty.  added integer slider and integer operations.
![image](https://github.com/pm-Architect/Verse3/assets/8115224/996032c5-a86e-475e-ab78-5c645a0f0fcc)


so.. project looks like:
![image](https://github.com/pm-Architect/Verse3/assets/8115224/15a6ca8b-51a7-4f43-8137-806720253b73)

before submitting for your approval, i would like to know your recommendation for ... allowing math of both integer and real values... your NUMBER vs the added INTEGER nodes and sliders.

i want to be able to do int to float... BUT!!! not without a performance warning as converting float to int or int to float can be expensive the CPU... blech.

is this acceptable?  how would you like me to post this.... for your review?  i don't want to merge as it has had condsiderable refactoring in preperation for extension.

I'd also like to begin diagraming your project if that is ok.  i use sparx EA (Enterprise Architect), it's awsome.   anyway...  i love what you've done.. still a lot of work to do... saving and loading... the compilation of csharp code.. but i see where you are headed and wish to ... join you for a bit.    ;-) but... if you allow, maybe i can be of assistance?

the problem being, when two creative minds join, do they make music like the blues?  i love the blues. 

anyway, if it's ok, can you make a branch of Verse3 where I can post my changes for your review and possible approval?  or at least a branch that you manage and i take in a direction that does not involve Blaze?  

up to your brother.  but i'd like you to see where i'm headed and if you would assist me in some refactoring that MUST preserve your business and data logic.  it's hard to refactor without loosing functionality.  AHHHHHHHHH ... can you tell? this weekend, i lost my mind, it fell out onto the ground.... and then i found it... smaller than before, but i found it.  LOL.

i had fun with your code.  and am still  pleased with your componenet plugin system.  facinating...  neat.

we should seperate your DataModel and formalize these classes as a form a Souce Code DSL.  a CST.
ps.  added Visitor Pattern to the Elements and DataModel in anticipation for walking the tree, generating dependecy and reduction graphs... and generating IML to maybe pass to LLVM for JIT (just in time) or for BIN (exe)...

anyyyyyway, back to reality and the real job.  let me know if i can post this for your review.

Thanks.

**Reply by RantNRave31** [reply]
Bool8->Integer, and Integer Slider, and a Binary Interleave (Morton Location Code).  dude.. this is kewl.  gonna add 3d interleave a quadtree and an octree test.
![image](https://github.com/pm-Architect/Verse3/assets/8115224/30c1106d-d746-4a37-b286-6927cc8c7d6c)

**Reply by RantNRave31** [reply]
Two way Binding Properties Dialog Added bound to DataModelView.SelectedElement
not all of your element viewModel are propigating up to the element modelView.
This means that two way binding on the property dialog is not relected back up correct chain to the model/view tree.
i.e. when I change the property in the properties window, the property of the element does change.  but the element modelView does not reflect as some Property do not have INotifyPropertyChanged.
![image](https://github.com/pm-Architect/Verse3/assets/8115224/a352f09b-8999-4978-a395-525d08ed5068)

**Reply by pm-Architect** [reply]
This is awesome!! Please write to me via email and lets discuss your plans with this going forward! You've got a good grip on the codebase clearly, so I'm curious to know where you see this project going. Preferably let's talk over a voice/video call to save time. I'm best available here: contact@praneetmathur.com

**Comment by RantNRave31** [comment]
Mr Mathur!
Nice to meet you.

mannnnnnnnnnnn..  this Verse3 is great.  just attempting to separate the
diagram elements from the data elements... or nodes?
I'm still learning your code and trying things out.

Need to add Library, and Plugin ViewModels for checking dependencies and
enabling/disabling (do you want loading unloading?) of the Assemblies.

Also, moved your Assembly to it's own namespace.



i'm having a hard time separating the underlying datamodel the arsenal and
the separate views for multi doc.

i think i'm going to need your assistance explaining some things.?

anyway, it'a bright and early day, and SGT Gunter here, (veteran, etc etc)
is up and highly motivated.

good morning.

;-)

I personally wish to see your Verse in and IDE and am working towards that
direction.
I wish to see your diagrams elements decoupled so i can use them for any
sort of diagram.

i.e class diagrams, that can provide "structures" for Verse3 components to
use.

so i suppose the first thing, is to get professional.  at 54, i'm no
concerned with looks, only results.

lets get pro.  ;-)  I you would start, with a few paragraphs on design
intent?  so that i don't unintentionally remove something you want in the
future.

this tool can be something quite like SSIS (SQL Server Integration
Services) where we create automations, like packages, and deploy them for
execution in a target environment.

this tool can be a lot of stuff depending on how we extend your plugin
framework.  outstanding job dude.

anyway, take a look at Visual Studio Modeling SDK, specifically
Dommain Specific Language Designers and "generating code"

yeah.  if we detangle model, modelview, and viewmodel, we can use different
models that verse under the "verse" diagrams elements who can host more
than .. function.

anyway, i'm crazy.  hope you can handle me.  most can't ... IQ over 135 but
no formal education.  33% hindi, 60% European, and some genetic odds and
ends from  Denisovans and Neanderthals.

A hard charging Engineer, not combat, .. you know, we build hospitals,
schools, roads, airfields, citys, hell, anything.

anyway, you can see that once i am out of well formed formula, EBNF
Grammars, etc, that i have no concern whatsoever with grammar?

yeah.  single line, one thought sentances.

hope it doesn't confuse you..  it's just a writting style called "stream of
conciousness"

;-)  oh, we are going to be adding Bayesian Networks very SOON!!!!

as you and i are hopefully, going to learn how to use your diagrams to
design models... with drag and drop features like CNN, classigication,
recognition, etc.

we are hopefully, if you allow, going to interface to OpenVDB (please read)
and MilVUS (long term AI memory storage, vectorized database)

hint hint.  the trick is licensing.. but you've got that started.  i'll add
a key, feature, issue thing for licensable features.

anywayyyyyyyy... you have a great day.. and if you get a chance, please
revfiew the Verse_1 fork and see if you can see a way to get me to multi
model views?

i'll leave "saving and loading" alone as your intent is to disable saving
for now.

no problem .  it's still fun.
I'll begin documenting and diagraming ... yeah.. that may help us
communicate.

Have a great day.

Don
